File: 03-Artificial-Neural-Networks/csharp-neural-network/Trainer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp_Neural_Network
{
    delegate bool CheckCorrect(double[] target, double[] output);

    class Trainer
    {
        readonly Network net;
        readonly Sample[] cases;
        readonly int caseCount;
        readonly CheckCorrect checkCorrect;
        readonly Sample[] testCases;

        public Trainer(Network net, Sample[] cases, 
            CheckCorrect checkCorrect, Sample[] testCases = null)
        {
            this.net = net;
            this.cases = cases;
            this.caseCount = cases.Length;

            if (caseCount == 0)
                throw new Exception("Don't have any samples to train with!");
            var sampleSample = cases[0];
            if (net.InputSize != sampleSample.input.Length)
                throw new Exception(string.Format(
                    "Net input size: {0}, doesn't match sample input size: {1}.",
                    net.InputSize, sampleSample.input.Length));
            if (net.OutputSize != sampleSample.target.Length)
                throw new Exception(string.Format(
                    "Net output size: {0}, doesn't match sample target size: {1}.",
                    net.OutputSize, sampleSample.target.Length));

            this.checkCorrect = checkCorrect;
            this.testCases = testCases ?? new Sample[0];
        }

        public long EpochSize { get { return caseCount; } }
        public long EpochCount { get; private set; }
        public long CurrentSampleCount { get; private set; }
        public long TotalSampleCount { get; private set; }

        public void TrainAllCases()
        {
            foreach (var sample in cases)
            {
                net.FeedForward(sample.input);
                net.PropagateBack(sample.target);
                CurrentSampleCount++;
                TotalSampleCount++;
            }
            EpochCount++;
            CurrentSampleCount = 0;
        }

        public void TrainUntilDone(bool showEpochStats = false)
        {
            var start = DateTime.UtcNow;
            Wr("Got {0:n0} training cases and {1:n0} test cases", caseCount, testCases.Length);
            var trainingAccuracy = MeasureAccuracy(cases);
            var testAccuracy = MeasureAccuracy(testCases);
            WriteAccuracy(trainingAccuracy, "Training");
            WriteAccuracy(testAccuracy, "    Test");
            Wr("Starting first epoch...");
            Wr("================================");
            while (trainingAccuracy.Item1 != trainingAccuracy.Item2)
            {
                var epochStart = DateTime.UtcNow;
                ShuffleCases();
                var trainingStart = DateTime.UtcNow;
                TrainAllCases();
                var trainingEnd = DateTime.UtcNow;
                trainingAccuracy = MeasureAccuracy(cases);
                if (showEpochStats || trainingAccuracy.Item1 == trainingAccuracy.Item2)
                {
                    Wr("Epoch {0} complete:", EpochCount);
                    testAccuracy = MeasureAccuracy(testCases);
                    WriteAccuracy(trainingAccuracy, "Training");
                    WriteAccuracy(testAccuracy, "    Test");
                    Wr("Training time: {0:mm\\:ss}", (trainingEnd - trainingStart));
                    Wr("   Epoch time: {0:mm\\:ss}", (DateTime.UtcNow - epochStart));
                    Wr("     Run time: {0:mm\\:ss}", (DateTime.UtcNow - start));
                    Wr("================================");
                }
            }
            Wr("All done.");
            Wr("");
        }

        static Random rnd = new Random();
        public void ShuffleCases()
        {
            for (int i = 0; i < caseCount; i++)
            {
                int r = (int)(rnd.NextDouble() * caseCount);
                var rTemp = cases[r];
                cases[r] = cases[i];
                cases[i] = rTemp;
            }
        }

        void WriteAccuracy(Tuple<int, int> t, string sampleName)
        {
            Wr("{0}: {1:n2}% correct ({2:n0}/{3:n0})",
                sampleName, GetPercent(t.Item1, t.Item2), t.Item1, t.Item2);
        }

        float GetPercent(int d, int n) { return 100 * (float)d / (float)n; }

        Tuple<int, int> MeasureAccuracy(Sample[] samples)
        {
            var correctCount = (
                from sample in samples
                let output = net.FeedForward(sample.input)
                where checkCorrect(sample.target, output)
                select 1)
                .Count();
            return Tuple.Create(correctCount, samples.Length);
        }

        void Wr(string text, params object[] args)
        {
            Console.WriteLine(text, args);
        }

    }
}

File: 03-Artificial-Neural-Networks/fsharp-neural-network2/Trainer.fs
module Trainer

open System
open Network
open Sample

type stats = 
  { mutable CurrentSampleCount : int64
    mutable TotalSampleCount : int64
    mutable EpochCount : int64 }

let trainAllCases net samples stats =
    samples 
    |> Array.iter(fun sample ->
        feedForward net sample.Input |> ignore
        propagateBack net sample.Target |> ignore
        stats.CurrentSampleCount <- stats.CurrentSampleCount + 1L
        stats.TotalSampleCount <- stats.TotalSampleCount + 1L )
    stats.EpochCount <- stats.EpochCount + 1L
    stats.CurrentSampleCount <- 0L

let measureAccuracy net samples checkCorrect =
    let correctCount =
        samples
        |> Seq.map (fun sample -> feedForward net sample.Input |> checkCorrect sample.Target)
        |> Seq.filter (id)
        |> Seq.length
    (correctCount, Array.length samples)

let writeAccuracy correctCount sampleCount sampleName =
    let getPercent n d = 100. * (float n) / (float d)
    Console.WriteLine("{0}: {1:n2}% correct ({2:n0}/{3:n0})",
        sampleName, (getPercent correctCount sampleCount), correctCount,sampleCount);

let rnd = new Random()
let shuffle cases =
    let caseCount = Array.length cases
    for i in seq{ 0 .. (caseCount - 1)} do
        let randomIndex = int (rnd.NextDouble() * float caseCount)
        let rTemp = cases.[randomIndex]
        cases.[randomIndex] <- cases.[i]
        cases.[i] <- rTemp

let trainUntilDone net (cases : Sample[]) (testCases : Sample[]) checkCorrect showEpochStats =
    let now() = DateTime.UtcNow
    let start = now()
    Console.WriteLine("Got {0:n0} training cases and {1:n0} test cases", cases.Length, testCases.Length);
    let mutable trainCorrect, trainCount = measureAccuracy net cases checkCorrect
    writeAccuracy trainCorrect trainCount "Training"
    let testCorrect, testCount = measureAccuracy net testCases checkCorrect
    writeAccuracy testCorrect testCount "    Test"
    Console.WriteLine("Starting first epoch...")
    Console.WriteLine("================================")
    let stats = {CurrentSampleCount = 0L; TotalSampleCount = 0L; EpochCount = 0L}
    while trainCorrect < trainCount do
        let epochStart = now()
        shuffle cases
        let trainingStart = now()
        let trainResult = trainAllCases net cases stats
        let trainingEnd = now()
        let newTrainCorrect, newTrainCount = measureAccuracy net cases checkCorrect
        trainCorrect <- newTrainCorrect
        trainCount <- newTrainCount
        if showEpochStats || trainCorrect = trainCount then
            Console.WriteLine("Epoch {0} complete:", stats.EpochCount)
            writeAccuracy trainCorrect trainCount "Training"
            let testCorrect, testCount = measureAccuracy net testCases checkCorrect
            writeAccuracy testCorrect testCount "    Test"
            Console.WriteLine("Training time: {0:mm\\:ss}", (trainingEnd - trainingStart))
            Console.WriteLine("   Epoch time: {0:mm\\:ss}", (DateTime.UtcNow - epochStart))
            Console.WriteLine("     Run time: {0:mm\\:ss}", (DateTime.UtcNow - start))
            Console.WriteLine("================================")
